First part of the revised docstrings

snilsn opened the pull request:
This PR contains the revision of the docstrings in _analysis, center_of_gravity, feature detection, segmentation, testing_ and _tracking_.

snilsn commented:
I was not able to figure out what the functions _cell_statistics_all(), cell_statistics()_ and _cog_cell()_ of _analysis.py_ do and how to use them. If anyone has used them I would be grateful for any advice.

mheikenfeld commented:
Hi Nils! I am probably the only one who can help with those functions in analysis.py. It is basically a set of analysis methods for the calculation of specific statistical properties along the cell tracks making use of the segmentation mask. I can have a look at making sure they still work, adding some more detailed documentation to them and putting together a little example notebooks over the next week or so!

snilsn commented:
Thanks Max, appreciate it!

freemansw1 commented:
Thanks Nils! I've requested @JuliaKukulies and I as reviewers for now, and I'll try to review this next week. Sorry for the delayed action, it's been busy for all of us. @mheikenfeld do you want to review as well, or would you rather your thoughts be more informal?

JuliaKukulies commented on tobac/segmentation.py:
```diff
@@ -11,6 +44,14 @@ def segmentation_3D(
     method="watershed",
     max_distance=None,
 ):
+    """Wraper for the segmentation()-function.
```
```suggestion
    """Wrapper for the segmentation()-function.
```

JuliaKukulies commented on tobac/testing.py:
```diff
@@ -79,21 +94,31 @@ def make_simple_sample_data_2D(data_type="iris"):
+    The grid has a grid spacing of 1km in both horizontal directions
+    and 100 grid cells in x direction and 200 in y direction.
+    Time resolution is 1 minute and the total length of the dataset is
+    100 minutes around a abritraty date (2000-01-01 12:00).
```
```suggestion
    100 minutes around a arbitrary date (2000-01-01 12:00).
```

JuliaKukulies reviewed (changes requested):
Many thanks to you @snilsn! Your great effort on the improved documentation is much appreciated! Before I approve, I have a few comments and suggestions, but most of them are only minor things or direct corrections of typos or misspellings

snilsn commented on tobac/analysis.py:
```diff
@@ -321,6 +568,34 @@ def calculate_velocity_individual(feature_old, feature_new, method_distance=None
+
+    Notes
+    -----
+    needs short summary, description and type of track
```
Yes, missed that

snilsn commented:
Thank you very much for your feedback, it is always interesting to see how small things can be easily missed, much appreciated. I fixed all the typos, deleted the unnecessary things and updated the reference.

On the other things I will either comment or work on within the next days.

snilsn commented on tobac/analysis.py:
```diff
@@ -19,6 +47,49 @@ def cell_statistics_all(
+    Parameters
+    ----------
+    input_cubes : iris.cube.Cube
+
```
Yes, I have just postponed working on this docstring until I can get the function to actually do something

snilsn commented on tobac/segmentation.py:
```diff
@@ -11,6 +44,14 @@ def segmentation_3D(
+    Notes
+    ----------
+
+    Obsolete?
```
I removed my comments and opened issue #145

snilsn commented on tobac/tracking.py:
```diff
@@ -24,31 +44,119 @@ def linking_trackpy(
+    pixel-based values used in trackpy. This allows for cloud tracking
+    that is controlled by physically-based parameters that are
+    independent of the temporal and spatial resolution of the input
+    data. The algorithm creates a continuous track for the cloud that
```
I'm not really sure what you mean here. Are you referring to the preceding sentence about the newly initialized trajectories?

snilsn replied:
That's a really interesting question, the answer seems to be non-trivial. 
The [documentation of trackpy](https://soft-matter.github.io/trackpy/dev/tutorial/subnets.html) refers to this [paper](https://crocker.seas.upenn.edu/CrockerGrier1996b.pdf) for the selection of the matching features, which is based on the probability for Brownian motion among the candidates. 

I've adjusted the wording to reflect this, still feels somewhat like a black box though.

JuliaKukulies replied:
All right, I was not sure whether this is something that I have missed and should now or if it is more complicated. Thanks for pointing out the reference and I think we can leave it with the description you have now.

snilsn commented on tobac/segmentation.py:
```diff
@@ -1,3 +1,36 @@
+cloud objects are directly connected, the border runs along the
+watershed line between the two regions. This procedure creates a mask of
+the same shape as the input data, with zeros at all grid points where
+there is no cloud or updraft and the integer number of the associated
```
You are right, this should be consistent and as general as possible. I think I will follow your suggestion and change `cloud` in the docstrings to `features` or `tracks`.

JuliaKukulies replied:
Very good, thanks @snilsn !

JuliaKukulies commented on tobac/feature_detection.py:
```diff
@@ -1,3 +1,23 @@
+"""Provide feature detection.
+
+This module can work with any two-dimensional field 
```
The full 3D implementation is currently under review and the next big thing to be implemented for v1.5, see also #127. We can also wait until this PR is merged and update the documentation then.

JuliaKukulies commented on tobac/analysis.py:
```diff
@@ -180,6 +295,35 @@ def cog_cell(
+
+    Notes
+    -----
+    Not sure what this function does
```
I am not sure what your view on this is @freemansw1 - should we wait until we have sorted this out or just leave out these docstrings for `analysis.py` and go ahead with the first bunch of improved documentation?

JuliaKukulies reviewed (commented):
Great, thank you for the quick response on that! I am happy with all the updates so far. Just waiting for our remaining discussion points and then @freemansw1 may want to have a look as well. Really great work :)

JuliaKukulies reviewed (approved):
I am happy with this now, many thanks @snilsn! The question is only if we want to wait until the missing docstrings in `analysis.py` are solved or not. I would almost suggest to go ahead without them, since @freemansw1 initiated a general discussion on future of these functions ( #146 ) and because this is only meant to be the first chunk of improved docstrings, not the complete revision. What is your opinion @freemansw1 ?

freemansw1 commented on tobac/analysis.py:
```diff
@@ -227,6 +371,46 @@ def cog_cell(
+    bin_centers : ndarray
+        The centers of the histogram intervalls.
+
+    minutes, optional : ndarray
```
I think it would be better here to specify which parameter controls whether this appears or not.

freemansw1 commented on tobac/analysis.py:
```diff
@@ -261,10 +459,31 @@ def haversine(lat1, lon1, lat2, lon2):
+    distance : float or pandas.Series
+        Float with the distance between the two features in meters if
+         the input are two pandas.Series containing one feature,
+         pandas.Series of the distancesif one of the inputs contains
```
small change- need a space between 'distances' and 'if'

freemansw1 commented on tobac/analysis.py:
```diff
@@ -563,6 +1044,56 @@ def area_histogram(
 def histogram_cellwise(
     Track, variable=None, bin_edges=None, quantity="max", density=False
 ):
+    """Create a histogram of the maximum, minimum or mean of
```
I think on this function and the corresponding `featurewise` function, it would be helpful in the documentation to include what we mean by featurewise and by cellwise.

freemansw1 commented on tobac/analysis.py:
```diff
@@ -579,6 +1110,46 @@ def histogram_cellwise(
 
 def histogram_featurewise(Track, variable=None, bin_edges=None, density=False):
+    """Create a histogram of a variable from the features of a
+    track. Essentially a wrapper of the numpy.histogram()
```
See comment above on cellwise

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -337,35 +409,54 @@ def feature_detection_multithreshold_timestep(
+        Threshold value used to select target regions to track. Default
+        is None.
+
+    min_num : int, optional
```
Agreed with @JuliaKukulies, but I do think we can go ahead and note that this parameter isn't used in this PR.

freemansw1 commented on tobac/feature_detection.py:
```diff
@@ -437,31 +526,61 @@ def feature_detection_multithreshold(
+    Returns
+    -------
+    features : pandas.DataFrame
+        Detected features.
```
I think this is a good idea.

freemansw1 commented on tobac/segmentation.py:
```diff
@@ -210,31 +287,64 @@ def segmentation(
-    Output:
-    segmentation_out: iris.cube.Cube
-                   Cloud mask, 0 outside and integer numbers according to track inside the cloud
+    """Use watershedding or random walker technique to determine region above
```
Random walker isn't supported in this implementation.

freemansw1 commented on tobac/segmentation.py:
```diff
@@ -286,10 +396,12 @@ def segmentation(
 
 
 def watershedding_3D(track, field_in, **kwargs):
+    """Wrapper for the segmentation()-function."""
```
These should probably also be depreciated; I will reopen #145 .

freemansw1 commented on tobac/tracking.py:
```diff
@@ -24,31 +45,124 @@ def linking_trackpy(
+        Maximum size of subnetwork for linking. This parameter should be
+        adjusted when using adaptive search. Usually a lower value is desired
+        in that case. For a more in depth explanation have look
+        `here <https://soft-matter.github.io/trackpy/v0.3.0/tutorial/adaptive-search.html>`_
```
This should link to their latest documentation rather than v0.3.0.

freemansw1 commented on tobac/tracking.py:
```diff
@@ -24,31 +45,124 @@ def linking_trackpy(
+    trajectories_final : xarray.Dataset
+        This enables filtering the resulting trajectories, e.g. to
+        reject trajectories that are only partially captured at the
+        boundaries of the input field both in space and time. [5]_
```
Where is this reference to?

snilsn replied:
I don't know, this is some fragment from older parts of the documentation I overlooked. I removed it and also changed this part for the tracking output to actually explain the resulting object.

freemansw1 reviewed (changes requested):
Overall, really excellent job. Just some minor changes here.

snilsn commented:
Thanks for your feedback, @freemansw1. This new update should address all the remarks you made, and also the remaining open conversations

JuliaKukulies commented on tobac/analysis.py:
```diff
@@ -19,6 +47,49 @@ def cell_statistics_all(
+
+    Notes
+    -----
+    Not sure what this function does
```
Yes I agree to remove the notes

snilsn replied:
The notes are removed now

freemansw1 reviewed (approved):
I'm happy with this now!

freemansw1 commented:
Actually I realized we should wait to merge this until we are staging for v1.4.0, and I am disinclined now from doing a squash merge, so I will plan on merging with a normal commit.

snilsn commented:
I would be happy with both options

JuliaKukulies commented:
I agree that it is good to wait for staging for v1.4.0 and not sure about the squash commit. I think @snilsn has written quite clear commit messages and kept the commits organized so that it is easy to follow what has been changed. It is just quite many commits, because my direct suggestions for typo fixes all come as single commits. Maybe I could find a way to only squash those? That would reduce the number of commits by more than half

freemansw1 commented:
As discussed in our meeting yesterday, I've moved this to target v1.4.0, so I'm now happy for us to merge this as long as you are happy @snilsn .

snilsn commented:
Yes, all good. Please go ahead
